### PR TITLE
Add Gardener upgrade tests

### LINKED ISSUE #2692: E2E test for deleting a shoot that wasn't reconciled with the new Gardener version yet
**How to categorize this issue?**

/area quality
/kind test
/priority normal

**What would you like to be added**:
We should add an integration test that tries to delete a `Shoot` that wasn't reconciled with the new Gardener versions yet.

**Why is this needed**:
Sometimes we add new functionality in the reconciliation flow that is also required by the deletion flow. However, not always the necessary code is also added to the deletion flow, i.e., when a shoot gets deleted without being reconciled first it might fail as certain assumptions don't hold true (see https://github.com/gardener/gardener/pull/2681).
An integration test for such case would be very helpful to detect such issues.

Disclaimer: In order to properly write such an integration test we need to be able to dynamically setup a new Gardener version. We agreed that we won't do it with the current garden-setup, thus, we'll first invest into a new deployment model with the landscaper (cc @dguendisch @schrodit) before tackling the mentioned integration test.

### PR DESCRIPTION
**Pull request by seshachalam-yv**
**How to categorize this PR?**
/area testing high-availability
/kind test enhancement

**What this PR does / why we need it**:

This PR adds below test cases to `Gardener` upgrade test suite.

- test to ensure that shoots which were hibernated in the previous release can successfully wake up and be deleted in the next Gardener version.

- test to ensure that non-HA shoots created in the previous Gardener release can be upgraded to HA and deleted in the next version of Gardener.

**Which issue(s) this PR fixes**:
Fixes #2692

**Special notes for your reviewer**:
cc: @rfranzke 

**Release note**:

```other operator

```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -452,10 +452,14 @@ test-e2e-local-operator: $(GINKGO)
 	./hack/test-e2e-local.sh operator --procs=1 --label-filter="default" ./test/e2e/operator/...
 
 test-pre-upgrade: $(GINKGO)
-	./hack/test-e2e-local.sh --procs=$(PARALLEL_E2E_TESTS) --label-filter="pre-upgrade" ./test/e2e/gardener/...
+	./hack/test-e2e-local.sh --procs=$(PARALLEL_E2E_TESTS) --label-filter="pre-upgrade && !high-availability" ./test/e2e/gardener/...
+test-ha-pre-upgrade: $(GINKGO)
+	./hack/test-e2e-local.sh --procs=$(PARALLEL_E2E_TESTS) --label-filter="pre-upgrade || (pre-upgrade && high-availability)" ./test/e2e/gardener/...
```
Comment (ScheererJ) [inline Makefile]: The label filter is confusing. The second part (`pre-upgrade && high-availability`) will never be relevant as the first part is always `true` if the second part is `true`.
This target executes all `pre-upgrade` tests, which does not match its name as the name implies something about high-availability.
Could you please explain?

File: Makefile [hunk 2]
```diff
@@ -452,10 +452,14 @@ test-e2e-local-operator: $(GINKGO)
...
 
 test-post-upgrade: $(GINKGO)
-	./hack/test-e2e-local.sh --procs=$(PARALLEL_E2E_TESTS) --label-filter="post-upgrade" ./test/e2e/gardener/...
+	./hack/test-e2e-local.sh --procs=$(PARALLEL_E2E_TESTS) --label-filter="post-upgrade && !high-availability" ./test/e2e/gardener/...
+test-ha-post-upgrade: $(GINKGO)
+	./hack/test-e2e-local.sh --procs=$(PARALLEL_E2E_TESTS) --label-filter="post-upgrade || (post-upgrade && high-availability)" ./test/e2e/gardener/...
```
Comment (ScheererJ) [inline Makefile]: Same as above.

**Comment by timuthy** [comment]
> should be able to restore etcd when etcd pod's pvc is corrupted in previous gardener release

Isn't this rather an e2e-test we want to have in Etcd-Druid or do you want to test any specific Gardener functionality here?

**Review by ScheererJ — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (seshachalam-yv) [reply Makefile]: ### **Label Filter:  "`pre-upgrade` || (`pre-upgrade` && `high-availability`)"**
For this label filter, the script will execute test cases with the `pre-upgrade` label or test cases with both "`pre-upgrade`" and `high-availability` labels:

- Test case 1: Will be executed (has `pre-upgrade` label) -- matched by the first part
- Test case 2: Will be executed (has `pre-upgrade` label) -- matched by the first part
- Test case 3: Will be executed (has both `pre-upgrade` and `high-availability` labels) -- matched by the **second part**
>  This means that all test cases with the "pre-upgrade" label will be executed, including those with the "high-availability" label.
<hr>

### **Label Filter:  "`pre-upgrade` && `!high-availability`)"**

For this label filter, the script will execute test cases with the `pre-upgrade` label  and not `high-availability` test cases.

- Test case 1: Will be executed (has `pre-upgrade` label) -- matched by the first part
- Test case 2: Will be executed (has `pre-upgrade` label) -- matched by the first part
- Test case 3: Will be executed (has both pre-upgrade and high-availability labels) -- not matched by the **second part**

<hr>

Comment (ScheererJ) [reply Makefile]: In boolean logic:
```
pre-upgrade || (pre-upgrade && high-availability) = pre-upgrade
```

Therefore, the high availability case is NOT the opposite of the ordinary case, but rather a superset.

Either the name is wrong or the expression does not do what you assume it does.

Comment (seshachalam-yv) [reply Makefile]: Yes, you are correct . Thanks @ScheererJ .

Fixed 
- `test-non-ha-pre-upgrade`:  test target executes `pre-upgrade` tests, excluding `high-availability`.

- `test-pre-upgrade`:  test target runs all `pre-upgrade` tests, including `high-availability`.

- `test-non-ha-post-upgrade`:  test target executes `post-upgrade` tests, excluding `high-availability`.

- `test-post-upgrade`:  test target runs all `post-upgrade` tests, including `high-availability`.

**Comment by seshachalam-yv** [comment]
> > should be able to restore etcd when etcd pod's pvc is corrupted in previous gardener release
> 
> Isn't this rather an e2e-test we want to have in Etcd-Druid or do you want to test any specific Gardener functionality here?

I understand your point, and it's indeed valuable to have such a test case in the Gardener E2E test suite. By including this test case, we can simulate real-time scenarios where an etcd PVC might get corrupted during the `Gardener` upgrade of a shoot cluster.

File: hack/ci-e2e-kind-upgrade.sh [hunk 3]
```diff
@@ -183,17 +205,17 @@ export GARDENER_PREVIOUS_VERSION="$(cat $GARDENER_RELEASE_DOWNLOAD_PATH/gardener
...
-trap '{
-  rm -rf "$GARDENER_RELEASE_DOWNLOAD_PATH/gardener-releases"
-  export_artifacts "$CLUSTER_NAME"
-  kind_down
-}' EXIT
+# trap '{
+#   rm -rf "$GARDENER_RELEASE_DOWNLOAD_PATH/gardener-releases"
+#   export_artifacts "$CLUSTER_NAME"
+#   kind_down
+# }' EXIT
```
Comment (ScheererJ) [inline hack/ci-e2e-kind-upgrade.sh]: This is probably a debugging leftover. Please revert this change.

File: hack/ci-e2e-kind-upgrade.sh [hunk 4]
```diff
@@ -212,6 +234,6 @@ TIMEOUT=1200 ./hack/usage/wait-for.sh seed "$SEED_NAME" GardenletReady Bootstrap
 sleep 60
 
 echo "Running gardener post-upgrade tests"
-make test-post-upgrade GARDENER_PREVIOUS_RELEASE=$GARDENER_PREVIOUS_RELEASE GARDENER_NEXT_RELEASE=$GARDENER_NEXT_RELEASE
+run_post_upgrade_test
 
-gardener_down
+# gardener_down
```
Comment (ScheererJ) [inline hack/ci-e2e-kind-upgrade.sh]: Same as above

**Review by ScheererJ — CHANGES_REQUESTED** [review]

**Comment by rfranzke** [comment]
> > > should be able to restore etcd when etcd pod's pvc is corrupted in previous gardener release
> > 
> > 
> > Isn't this rather an e2e-test we want to have in Etcd-Druid or do you want to test any specific Gardener functionality here?
> 
> I understand your point, and it's indeed valuable to have such a test case in the Gardener E2E test suite. By including this test case, we can simulate real-time scenarios where an etcd PVC might get corrupted during the `Gardener` upgrade of a shoot cluster.

I have trouble understanding this scenario-  why would an etcd PVC get corrupt during a Gardener upgrade? Can you explain when/how this would happen?
From my current understanding, I would rather second @timuthy's point and agree that such test case rather belongs to `gardener/etcd-druid` and is not related to `gardener/gardener`.

**Comment by rfranzke** [comment]
> **Gardener upgrade Tests for Shoot::e2e-upgrade-ha when `Pre-upgrade` (version:`'v1.61.4'`)**
> 
> * should create a shoot [high-availability, pre-upgrade]
> 
> **Gardener upgrade Tests for Shoot::e2e-upgrade when `Post-Upgrade` (Gardener version:`'v1.62.0-dev'`)**
> 
> * should be able to upgrade a non-HA shoot which was created in previous gardener release to HA with failure tolerance type 'zone' [high-availability, post-upgrade]
> * should be able to delete a shoot which was created in previous gardener release [high-availability, post-upgrade]

What is this test case about? We upgrade the non-HA shoot to HA after we upgraded Gardener, and then delete it? Don't we already have upgrade tests like this? In https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/7786/pull-gardener-e2e-kind-ha-multi-zone-upgrade/1646807961093279744, I see

> Test E2E Gardener Suite: [It] Gardener upgrade Tests for Shoot::e2e-upgrade-ha when Post-Upgrade (Gardener version:'v1.69.0-dev', Git version:'v1.69.0-dev') should be able to upgrade a non-HA shoot which was created in previous gardener release to HA with failure tolerance type 'zone' [high-availability, post-upgrade]

Isn't this exactly this case?

Also, what about https://github.com/gardener/gardener/issues/2692? Isn't such scenario (deleting without previous reconciliation) more valuable?

File: hack/ci-e2e-kind-upgrade.sh (hunk 3 above)
Comment (seshachalam-yv) [reply hack/ci-e2e-kind-upgrade.sh]: Thanks @ScheererJ 
Fixed

**Comment by seshachalam-yv** [comment]
> What is this test case about? We upgrade the non-HA shoot to HA after we upgraded Gardener, and then delete it? Don't we already have upgrade tests like this? In https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/7786/pull-gardener-e2e-kind-ha-multi-zone-upgrade/1646807961093279744, I see

In this job `pull-gardener-e2e-kind-ha-multi-zone-upgrade` we are trying to upgrade `non-HA` shoot cluster to `HA` in same release.

In this test case, the goal is to upgrade a `non-HA` shoot created in the previous release to an `HA` shoot in the next release. We discovered a critical bug https://github.com/gardener/gardener/pull/7434 while I was testing this test case only.

**Comment by seshachalam-yv** [comment]
> I have trouble understanding this scenario- why would an etcd PVC get corrupt during a Gardener upgrade? Can you explain when/how this would happen?

@rfranzke Let's say, for some reason, `etcd's` PVC gets corrupted during or before the Gardener upgrade. The next release should not be affected in terms of restoring the data. This test case is not only about PVC corruption; it also covers scenarios where one of the etcd nodes is down and recovering. There should not be any downtime in the cluster, and this test ensures that the next release does not cause any downtime while one of the `etcd` nodes is recovering.

**Comment by seshachalam-yv** [comment]
> Isn't this exactly this case?
> 
> Also, what about #2692? Isn't such scenario (deleting without previous reconciliation) more valuable?

We are already covering this test case:
> should able to delete a shoot which was created in previous release [post-upgrade]

This test case addresses the scenario you mentioned (#2692), ensuring that a shoot created in the previous release can be deleted successfully

**Comment by rfranzke** [comment]
> > I have trouble understanding this scenario- why would an etcd PVC get corrupt during a Gardener upgrade? Can you explain when/how this would happen?
> 
> This test case is not only about PVC corruption; it also covers scenarios where one of the etcd nodes is down and recovering.

Again, even if it's about this, this is all something related to etcd-druid, isn't it? Which role does Gardener play here? It has nothing to do with managing the etcd pods other than deploying the desired state of the `Etcd` resource. I still think all this belongs to etcd-druid only.

**Comment by rfranzke** [comment]
> > Isn't this exactly this case?
> > Also, what about #2692? Isn't such scenario (deleting without previous reconciliation) more valuable?
> 
> We are already covering this test case:
> 
> > should able to delete a shoot which was created in previous release [post-upgrade]
> 
> This test case addresses the scenario you mentioned (#2692), ensuring that a shoot created in the previous release can be deleted successfully

OK, then you can probably update the PR description with "fixes #2692" so that this issue gets closed when this PR is merged?

**Comment by seshachalam-yv** [comment]
> Again, even if it's about this, this is all something related to etcd-druid, isn't it? Which role does Gardener play here? It has nothing to do with managing the etcd pods other than deploying the desired state of the `Etcd` resource. I still think all this belongs to etcd-druid only.

Yes, removed `ETCD` related test cases.

**Comment by seshachalam-yv** [comment]
@rfranzke @timuthy @ScheererJ 
The PR is ready for review. I have removed the ETCD-related test cases and updated the PR description.

File: hack/ci-e2e-kind-upgrade.sh [hunk 5]
```diff
@@ -170,6 +170,28 @@ function set_seed_name() {
...
+function run_post_upgrade_test() {
+  case "$SHOOT_FAILURE_TOLERANCE_TYPE" in
+  node | zone)
+    make test-post-upgrade GARDENER_PREVIOUS_RELEASE=$GARDENER_PREVIOUS_RELEASE GARDENER_NEXT_RELEASE=$GARDENER_NEXT_RELEASE
+    ;;
+  *)
+    make test-non-ha-post-upgrade GARDENER_PREVIOUS_RELEASE=$GARDENER_PREVIOUS_RELEASE GARDENER_NEXT_RELEASE=$GARDENER_NEXT_RELEASE
+    ;;
+  esac
+}
```
Comment (ScheererJ) [inline hack/ci-e2e-kind-upgrade.sh]: The functions duplicate the make calls and the parameters. It is fairly verbose and explicit. You could also set a variable in case failure tolerance type is set and then just call something like `make test${prefix}-post-upgrade ...`.
It would be somewhat shorter in terms of lines of code and have less duplication, but may be a bit harder to read.
I have no hard opinion, which would be the better approach.

File: test/e2e/gardener/upgrade/upgrade.go [hunk 6]
```diff
@@ -121,4 +118,133 @@ var _ = Describe("Gardener upgrade Tests for", func() {
...
+		})
+	})
+
+	Context("Shoot::e2e-upgrade-hib", func() {
+		var (
+			parentCtx = context.Background()
+			f         = framework.NewShootCreationFramework(&framework.ShootCreationConfig{
+				GardenerConfig: e2e.DefaultGardenConfig(projectNamespace),
+			})
+			shootTest = e2e.DefaultShoot("e2e-upgrade-hib")
```
Comment (ScheererJ) [inline test/e2e/gardener/upgrade/upgrade.go:184]: Is the abbreviation necessary?

File: hack/test-e2e-local.sh [hunk 7]
```diff
@@ -53,11 +53,13 @@ if [[ "$1" != "operator" ]]; then
     e2e-update-node.local
     e2e-update-zone.local
     e2e-upgrade.local
+    e2e-upgrade-ha.local
+    e2e-upgrade-hib.local
```
Comment (ScheererJ) [inline hack/test-e2e-local.sh:57]: Is the abbreviation necessary?

**Review by ScheererJ — COMMENTED** [review]
Thanks. Looks good overall.

File: hack/test-e2e-local.sh (hunk 7 above)
Comment (seshachalam-yv) [reply hack/test-e2e-local.sh:57]: There is a limitation of 15 characters for shoot name, and I followed the same convention to indicate that it is an E2e upgrade test case for the hibernation scenario.

File: test/e2e/gardener/upgrade/upgrade.go (hunk 6 above)
Comment (seshachalam-yv) [reply test/e2e/gardener/upgrade/upgrade.go:184]: There is a limitation of 15 characters for shoot name, and I followed the same convention to indicate that it is an E2e upgrade test case for the hibernation scenario.

File: hack/ci-e2e-kind-upgrade.sh (hunk 5 above)
Comment (seshachalam-yv) [reply hack/ci-e2e-kind-upgrade.sh]: Fixed

File: test/e2e/gardener/upgrade/upgrade.go [hunk 8]
```diff
@@ -121,4 +118,133 @@ var _ = Describe("Gardener upgrade Tests for", func() {
...
+			BeforeAll(func() {
+				ctx, cancel = context.WithTimeout(parentCtx, 60*time.Minute)
+				DeferCleanup(cancel)
+				Expect(f.GetShoot(ctx, shootTest)).To(Succeed())
+				f.ShootFramework, err = f.NewShootFramework(ctx, shootTest)
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			It("should be able to upgrade a non-HA shoot which was created in previous gardener release to HA with failure tolerance type '"+
+				os.Getenv("SHOOT_FAILURE_TOLERANCE_TYPE")+"'", func() {
```
Comment (rfranzke) [inline test/e2e/gardener/upgrade/upgrade.go]: put this in the same line for readability

**Review by rfranzke — COMMENTED** [review]
Thanks, looks quite good now :)

File: test/e2e/gardener/upgrade/upgrade.go (hunk 8 above)
Comment (seshachalam-yv) [reply test/e2e/gardener/upgrade/upgrade.go]: 👍 Fixed.

File: test/e2e/gardener/common.go [hunk 9]
```diff
@@ -116,3 +119,19 @@ func DefaultShoot(name string) *gardencorev1beta1.Shoot {
 		},
 	}
 }
+
+// SetupDNSForMultiZoneTest sets the golang DefaultResolver to the CoreDNS server, which is portforwarded to the host 127.0.0.1:5353.
```
Comment (acumino) [inline test/e2e/gardener/common.go]: Nit
```suggestion
// SetupDNSForMultiZoneTest sets the golang DefaultResolver to the CoreDNS server, which is port forwarded to the host 127.0.0.1:5353.
```

File: test/framework/gardener_utils.go [hunk 10]
```diff
@@ -486,3 +491,28 @@ func setHibernation(shoot *gardencorev1beta1.Shoot, hibernated bool) {
...
+// VerifyNoRunningPods verifies that no control plane pods are running for a given shoot.
+// If any control plane pods are found to be running, returns an error with their names. Otherwise, returns nil.
+func (f *GardenerFramework) VerifyNoRunningPods(ctx context.Context, shoot *gardencorev1beta1.Shoot) error {
+	_, seedClient, err := f.GetSeed(ctx, *shoot.Spec.SeedName)
+	if err != nil {
+		return err
+	}
+
+	shootSeedNamespace := shoot.Status.TechnicalID
+	podList := &corev1.PodList{}
```
Comment (acumino) [inline test/framework/gardener_utils.go]: You can use metadata only list here.

Comment (seshachalam-yv) [reply test/framework/gardener_utils.go]: Fixed

**Review by acumino — COMMENTED** [review]
/lgtm
